### DISCUSSION TITLE
What's the purpose of adding cprover library?

### DISCUSSION THREAD

**Discussion by kunjsong01** [esbmc/esbmc#726] [Q&A] [answered]
Trying to understand the steps in `clang_c_languaget:final` funciton, but I'm not quite following the code in `add_cproveR_library`:https://github.com/esbmc/esbmc/blob/74b1ce7cd6ecca91a2012f105f821a50f3130303/src/clang-c-frontend/clang_c_language.cpp#L278-L282

### Questions:
Q1. Which cproverl library is it referring to? 
Q2. `clib32_buf`, `clib32_buf_size` - what are they referring to? 
https://github.com/esbmc/esbmc/blob/74b1ce7cd6ecca91a2012f105f821a50f3130303/src/c2goto/cprover_library.cpp#L19-L29
Q3. I don't understand the comment in `read_goto_binary.cpp` - Isn't GOTO program generated by `goto_convert` converting `symbolt` to `goto_programt`? 
https://github.com/esbmc/esbmc/blob/74b1ce7cd6ecca91a2012f105f821a50f3130303/src/goto-programs/read_goto_binary.cpp#L3

**Comment by fbrausse** [comment, marked as answer]
There is a CMake switch `ESBMC_BUNDLE_LIBC` enabled by default and used in https://github.com/esbmc/esbmc/blob/74b1ce7cd6ecca91a2012f105f821a50f3130303/src/c2goto/CMakeLists.txt#L27-L31 to precompile the operational models in https://github.com/esbmc/esbmc/tree/master/src/c2goto/library; this is the "cprover library", which I sometimes refer to this as "internal" or "ESBMC's libc", see also the output after starting ESBMC:
```
ESBMC version 6.9.0 64-bit x86_64 linux
Target: 64-bit little-endian x86_64-unknown-linux with esbmc libc
```
The binary `src/c2goto/c2goto` is used for as compiler that purpose. It outputs a representation of the GOTO code including the irep expressions for the functions in the internal libc in a custom binary format, which we talked about a bit in #557.

During standard compilation of ESBMC, the ilibc binaries, each corresponding to a different set of command-line switches (see above), are included verbatim in .o files generated by the build process from generated .c sources via https://github.com/esbmc/esbmc/blob/74b1ce7cd6ecca91a2012f105f821a50f3130303/scripts/cmake/Utils.cmake#L35. These generated sources look like
```c
const char clib32_buf[] = {
71,66,70,0,0,0,1,0,0,6,109,0,0,0,0,0,
0,0,0,0,78,0,0,0,27,116,121,112,101,0,0,0,
0,1,0,0,0,146,115,105,103,110,101,100,98,118,0,78,
/* [...] */
};
const unsigned int clib32_buf_size = sizeof(clib32_buf);
```
and can be found in `build/src/c2goto/clib*.c`. These precompiled GOTO binaries are read by the module you cited.

In addition to these precompiled versions of the libc, also the libc sources are bundled and included verbatim, in case the set of command-line flags ESBMC is run with does not correspond to a precompiled version. In that case `add_bundled_library_sources()` is run, which translates the C code of the internal libc at runtime (a pretty time consuming process): https://github.com/esbmc/esbmc/blob/74b1ce7cd6ecca91a2012f105f821a50f3130303/src/c2goto/cprover_library.cpp#L160-L166.

**Reply by kunjsong01** [reply]
Understood. Thanks for the answers!
